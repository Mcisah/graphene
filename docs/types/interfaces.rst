Interfaces
==========

An Interface contains the essential fields that will be implemented by
multiple ObjectTypes.

The basics:

- Each Interface is class decorated with ``InterfaceType``.
- Each attribute decorated with ``@Field`` represents a GraphQL Field in the
  Interface.

Quick example
-------------

This example model defines a ``Character`` interface with a name. ``Human``
and ``Droid`` are two implementations of that interface.

.. code:: js

    import { InterfaceType, ObjectType } from "graphene-js";

    @InterfaceType()
    class Character {
        @Field(String) name;
    }

    // Human is a Character implementation
    @ObjectType({
        interfaces: [Character]
    })
    class Human {
        @Field(String) bornIn;
    }

    // Droid is a Character implementation
    @ObjectType({
        interfaces: [Character]
    })
    class Human {
        @Field(String) function;
    }


``name`` is a field on the ``Character`` interface that will also exist on both
the ``Human`` and ``Droid`` ObjectTypes (as those implement the ``Character``
interface). Each ObjectType may define additional fields.

The above types have the following representation in a schema:

.. code::

    interface Character {
      name: String
    }

    type Droid implements Character {
      name: String
      function: String
    }

    type Human implements Character {
      name: String
      bornIn: String
    }
