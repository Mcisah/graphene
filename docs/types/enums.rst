Enums
=====

A ``Enum`` is a special ``GraphQL`` type that represents a set of
symbolic names (members) bound to unique, constant values.

Definition
----------

You can create an ``Enum`` using classes:

.. code:: js

    import { EnumType } from "graphene";

    @EnumType()
    class Episode {
        static NEWHOPE = 4
        static EMPIRE = 5
        static JEDI = 6
    }


Graphene will automatically search for the static variables in the Enum and expose
them as the enum values.

Value descriptions
------------------

It's possible to add a description to an enum value, for that the enum value
needs to have the ``description``decorator on it.

.. code:: js

    @EnumType()
    class Episode {
        @description("New hope episode")
        static NEWHOPE = 4

        @description("Empire episode")
        static EMPIRE = 5

        @description("JEDI episode")
        static JEDI = 6
    }
