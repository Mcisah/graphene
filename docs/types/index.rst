===============
Types Reference
===============

.. toctree::
   :maxdepth: 1

   enums
   scalars
   list-and-nonnull
   interfaces
   unions
   objecttypes
   schema
