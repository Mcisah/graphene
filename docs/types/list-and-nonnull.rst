Lists and Non-Null
==================

Object types, scalars, and enums are the only kinds of types you can
define in Graphene. But when you use the types in other parts of the
schema, or in your query variable declarations, you can apply additional
type modifiers that affect validation of those values.


List
----

.. code:: js

    import { ObjectType, List } from "graphene";

    @ObjectType()
    class Character {
        @Field(List(String)) appearsIn;
    }

Lists work in a similar way: We can use a type modifier to mark a type as a
``List``, which indicates that this field will return a list of that type.
It works the same for arguments, where the validation step will expect a list
for that value.

For ease of development, we can directly use js lists with one element ``[]``.

Like:


.. code:: js

    import { ObjectType } from "graphene";

    @ObjectType()
    class Character {
        @Field([String]) appearsIn;
    }


NonNull
-------

.. code:: js

    import { ObjectType, NonNull } from "graphene";

    @ObjectType()
    class Character {
        @Field(NonNull(String)) name;
    }


Here, we're using a ``String`` type and marking it as Non-Null by wrapping
it using the ``NonNull`` class. This means that our server always expects
to return a non-null value for this field, and if it ends up getting a
null value that will actually trigger a GraphQL execution error,
letting the client know that something has gone wrong.
