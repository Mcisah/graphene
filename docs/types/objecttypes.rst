ObjectTypes
===========

An ObjectType is the single, definitive source of information about your
data. It contains the essential fields and behaviors of the data youâ€™re
querying.

The basics:

- Each ObjectType is a Python class that inherits from
  ``graphene.ObjectType``.
- Each attribute of the ObjectType represents a ``Field``.

Quick example
-------------

This example model defines a Person, with a first and a last name:

.. code:: js

    import { ObjectType } from "graphene";

    @ObjectType()
    class Person {
        @Field(String) firstName;
        @Field(String) lastName;
        @Field(String)
        fullName() {
            return `${this.firstName} ${this.lastName}`;
        }
    }

**firstName** and **lastName** are fields of the ObjectType. Each
field is specified as a class attribute, and each attribute maps to a
Field.

The above ``Person`` ObjectType has the following schema representation:

.. code::

    type Person {
      firstName: String
      lastName: String
      fullName: String
    }


Resolvers
---------

A resolver is a method that resolves certain fields within a
``ObjectType``. If not specififed otherwise, the resolver of a
field is the ``resolve_{field_name}`` method on the ``ObjectType``.

By default resolvers take the arguments ``args``, ``context`` and ``info``.


Quick example
~~~~~~~~~~~~~

This example model defines a ``Query`` type, which has a reverse field
that reverses the given ``word`` argument using the ``resolve_reverse``
method in the class.

.. code:: js

    import { ObjectType } from "graphene";

    @ObjectType()
    class Query {
        @Field(String, {args: {word: String}})
        reverse({word}) {
            return (word || "").split("").reverse().join("")
        }
    }


Instances as data containers
----------------------------

Graphene ``ObjectType``\ s can act as containers too. So with the
previous example you could do:

.. code:: js

    peter = new Person({firstName: "", lastName: ""})

    peter.firstName # prints "Peter"
    peter.lastName # prints "Griffin"

.. _Interface: /docs/interfaces/
