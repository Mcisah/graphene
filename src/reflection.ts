/**
 * Copyright (c) 2017-present, Graphene.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import 'reflect-metadata';
import {
  GraphQLType,
  GraphQLString,
  GraphQLFloat,
  GraphQLList,
  GraphQLBoolean,
  assertType,
  isType,
  GraphQLFieldConfig,
  GraphQLInputFieldConfig
} from 'graphql';
import { GraphQLDateTime } from 'graphql-iso-date';

export const GRAPHENE_TYPE_METADATA_KEY = 'graphene:type';
export const GRAPHENE_FIELDS_METADATA_KEY = 'graphene:fields';
export const GRAPHENE_INPUTFIELDS_METADATA_KEY = 'graphene:inputfields';
export const GRAPHENE_DESCRIPTION_KEY = 'graphene:description';
export const GRAPHENE_DEPRECATED_KEY = 'graphene:deprecated';

// A utility funciton to convert values like:
//   [String]
// To:
//   GraphQLList(GraphQLString)
export const convertArrayToGraphQLList = (
  target: Array<any>
): GraphQLList<GraphQLType> => {
  if (target.length !== 1) {
    throw new Error(
      `Graphene Array definitions should contain exactly one element. Received ${
        target.length
      } elements.`
    );
  }
  var innerType = getGraphQLType(target[0]);
  return new GraphQLList(innerType);
};

// A utility funciton to convert values like:
//   String
// To:
//   GraphQLString
// Note: for String -> GraphQLString conversion to work we have to call
// setupNativeTypes
export const getGraphQLType = (target: any): GraphQLType => {
  if (Array.isArray(target)) {
    return convertArrayToGraphQLList(target);
  }
  if (isType(target)) {
    return target;
  }
  var metadataType: GraphQLType = Reflect.getMetadata(
    GRAPHENE_TYPE_METADATA_KEY,
    target
  );
  if (metadataType) {
    return metadataType;
  }
  throw new Error(
    `The target ${target} have no GraphQL type associated to it.`
  );
};

// An utility function to associate a GraphQL type
// with the specified target.
// This method takes full advantage of the Reflection API.
export const setGraphQLType = (target: any, type: GraphQLType): void => {
  // Is the provided type a GraphQLType?
  // Will fail if not.
  assertType(type);

  // First we check this type have no other type associated with it.
  if (Reflect.hasMetadata(GRAPHENE_TYPE_METADATA_KEY, target)) {
    throw new Error(
      `Type ${String(target)} already have a Graphene type attached.`
    );
  }
  // We define the type metadata through reflection
  Reflect.defineMetadata(GRAPHENE_TYPE_METADATA_KEY, type, target);
};

// Field definitions
export type UnmountedFieldMap = {
  [key: string]: () => GraphQLFieldConfig<any, any>;
};

export type MountedFieldMap = {
  [key: string]: GraphQLFieldConfig<any, any>;
};

// Get the fields given a constructor
export const getFields = (target: any): UnmountedFieldMap => {
  var fields: UnmountedFieldMap;
  if (!Reflect.hasMetadata(GRAPHENE_FIELDS_METADATA_KEY, target)) {
    fields = {};
    Reflect.defineMetadata(GRAPHENE_FIELDS_METADATA_KEY, fields, target);
  } else {
    fields = Reflect.getMetadata(GRAPHENE_FIELDS_METADATA_KEY, target);
  }
  return fields;
};

// mountFields
export const mountFields = (
  fields: UnmountedFieldMap
) => (): MountedFieldMap => {
  var key: string;
  var finalFields: MountedFieldMap = {};
  for (key in fields) {
    finalFields[key] = fields[key]();
  }
  return finalFields;
};

export const assertFields = (target: any, fields: UnmountedFieldMap) => {
  if (Object.keys(fields).length === 0) {
    throw new Error(`Type ${target} must have fields defined on it.`);
  }
};

// InputFieldDefinitions
export type UnmountedInputFieldMap = {
  [key: string]: () => GraphQLInputFieldConfig;
};

export type MountedInputFieldMap = {
  [key: string]: GraphQLInputFieldConfig;
};

// Get the fields given a constructor
export const getInputFields = (target: any): UnmountedInputFieldMap => {
  var fields: UnmountedInputFieldMap;
  if (!Reflect.hasMetadata(GRAPHENE_INPUTFIELDS_METADATA_KEY, target)) {
    fields = {};
    Reflect.defineMetadata(GRAPHENE_INPUTFIELDS_METADATA_KEY, fields, target);
  } else {
    fields = Reflect.getMetadata(GRAPHENE_INPUTFIELDS_METADATA_KEY, target);
  }
  return fields;
};

// mountInputFields
export const mountInputFields = (
  fields: UnmountedInputFieldMap
) => (): MountedInputFieldMap => {
  var key: string;
  var finalFields: MountedInputFieldMap = {};
  for (key in fields) {
    finalFields[key] = fields[key]();
  }
  return finalFields;
};

export const assertInputFields = (
  target: any,
  fields: UnmountedInputFieldMap
) => {
  if (Object.keys(fields).length === 0) {
    throw new Error(`Type ${target} must have fields defined on it.`);
  }
};

// The setup function that permit us to use
// the native String and Number types.
export const setupNativeTypes = () => {
  setGraphQLType(String, GraphQLString);
  setGraphQLType(Boolean, GraphQLBoolean);
  setGraphQLType(Number, GraphQLFloat);
  setGraphQLType(Date, GraphQLDateTime);
};

// Description setter
// Usage:
//   setDescription(MyClass, 'the description')
// Or:
//   setDescription(MyClass, 'methodName', 'the description')
export const setDescription = (
  target: any,
  keyOrDescription: string,
  description?: string
) => {
  if (typeof description === 'undefined') {
    Reflect.defineMetadata(GRAPHENE_DESCRIPTION_KEY, keyOrDescription, target);
  } else {
    Reflect.defineMetadata(
      GRAPHENE_DESCRIPTION_KEY,
      description,
      target,
      keyOrDescription
    );
  }
};

// Description getter
// Usage:
//   getDescription(MyClass)
// Or:
//   setDescription(MyClass, 'methodName')
export const getDescription = (target: any, key?: string) => {
  if (key) {
    return Reflect.getMetadata(GRAPHENE_DESCRIPTION_KEY, target, key);
  }
  return Reflect.getMetadata(GRAPHENE_DESCRIPTION_KEY, target);
};

// Deprecation reason setter
// Usage:
//   setDeprecationReason(MyClass, 'methodName', 'this method is now deprecated')
export const setDeprecationReason = (
  target: any,
  key: string,
  reason: string
) => {
  Reflect.defineMetadata(GRAPHENE_DEPRECATED_KEY, reason, target, key);
};

// Deprecation reason getter
// Usage:
//   getDeprecationReason(MyClass, 'methodName')
export const getDeprecationReason = (target: any, key: string) => {
  return Reflect.getMetadata(GRAPHENE_DEPRECATED_KEY, target, key);
};

// Setup the decorators

// Description decorator.
// This decorator permits us to decorate classes, like:
//
//   @description('The class description')
//   class X {}
//
// Or the class methods, like:
//
//   class X{
//     @description('The method description')
//     myMethod() {}
//   }
export const description = (description: string) => (
  target: any,
  key?: string,
  descriptor?: PropertyDescriptor
) => {
  if (typeof key !== 'undefined') {
    // It's a decorated method
    setDescription(target, key, description);
    return descriptor;
  }
  // It's a decorated class
  setDescription(target, description);
  return target;
};

// This decorator permits us to mark fields as deprecated, like:
//
//   class X{
//     @deprecated('This method is deprecated')
//     myMethod() {}
//   }
export const deprecated = (reason: string) => <MethodDecorator>(
  target: any,
  key?: string,
  descriptor?: any
) => {
  if (typeof key !== 'undefined') {
    // It's a decorated method
    setDeprecationReason(target, key, reason);
    return descriptor;
  }
  // It's a decorated class
  throw new Error(
    `Classes can't be decorated with the @deprecated decorator. Received ${target}.`
  );
};
