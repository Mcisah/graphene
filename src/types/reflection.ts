/**
 * Copyright (c) 2017-present, Graphene.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import 'reflect-metadata';
import {
  GraphQLType,
  GraphQLString,
  GraphQLFloat,
  GraphQLList,
  assertType,
  isType
} from 'graphql';

export const GRAPHENE_TYPE_METADATA_KEY = 'graphene:type';
export const GRAPHENE_FIELDS_METADATA_KEY = 'graphene:fields';
export const GRAPHENE_DESCRIPTION_KEY = 'graphene:description';
export const GRAPHENE_DEPRECATED_KEY = 'graphene:deprecated';
// export const GRAPHENE_ARGUMENTS_METADATA_KEY = 'graphene:arguments';

// A utility funciton to convert values like:
// [String]
// To:
// GraphQLList(GraphQLString)
export const convertArrayToGraphQLList = (
  target: Array<any>
): GraphQLList<GraphQLType> => {
  if (target.length !== 1) {
    throw new Error(
      `Graphene Array definitions should contain exactly one element. Received ${
        target.length
      } elements.`
    );
  }
  var innerType = getGraphQLType(target[0]);
  return new GraphQLList(innerType);
};

// A utility funciton to convert values like:
// String
// To:
// GraphQLString
// Note: for String -> GraphQLString conversion to work we have to call
// setupNativeTypes
export const getGraphQLType = (target: any): GraphQLType => {
  if (Array.isArray(target)) {
    return convertArrayToGraphQLList(target);
  }
  if (isType(target)) {
    return target;
  }
  var metadataType: GraphQLType = Reflect.getMetadata(
    GRAPHENE_TYPE_METADATA_KEY,
    target
  );
  if (metadataType) {
    return metadataType;
  }
  throw new Error(
    `The target ${target} have no GraphQL type associated to it.`
  );
};

// An utility function to associate a GraphQL type
// with the specified target.
// This method takes full advantage of the Reflection API.
export const setGraphQLType = (target: any, type: GraphQLType): void => {
  // Is the provided type a GraphQLType?
  // Will fail if not.
  assertType(type);

  // First we check this type have no other type associated with it.
  if (Reflect.hasMetadata(GRAPHENE_TYPE_METADATA_KEY, target)) {
    throw new Error(
      `Type ${String(target)} already have a Graphene type attached.`
    );
  }
  // We define the type metadata through reflection
  Reflect.defineMetadata(GRAPHENE_TYPE_METADATA_KEY, type, target);
};

// The setup function that permit us to use
// the native String and Number types.
export const setupNativeTypes = () => {
  setGraphQLType(String, GraphQLString);
  setGraphQLType(Number, GraphQLFloat);
};

// Description setter
// Usage:
//   setDescription(MyClass, 'the description')
// Or:
//   setDescription(MyClass, 'methodName', 'the description')
export const setDescription = (
  target: any,
  keyOrDescription: string,
  description?: string
) => {
  if (typeof description === 'undefined') {
    Reflect.defineMetadata(GRAPHENE_DESCRIPTION_KEY, keyOrDescription, target);
  } else {
    Reflect.defineMetadata(
      GRAPHENE_DESCRIPTION_KEY,
      description,
      target,
      keyOrDescription
    );
  }
};

// Description getter
// Usage:
//   getDescription(MyClass)
// Or:
//   setDescription(MyClass, 'methodName')
export const getDescription = (target: any, key?: string) => {
  if (key) {
    return Reflect.getMetadata(GRAPHENE_DESCRIPTION_KEY, target, key);
  }
  return Reflect.getMetadata(GRAPHENE_DESCRIPTION_KEY, target);
};

// Deprecation reason setter
// Usage:
//   setDeprecationReason(MyClass, 'methodName', 'this method is now deprecated')
export const setDeprecationReason = (
  target: any,
  key: string,
  reason: string
) => {
  Reflect.defineMetadata(GRAPHENE_DEPRECATED_KEY, reason, target, key);
};

// Deprecation reason getter
// Usage:
//   getDeprecationReason(MyClass, 'methodName')
export const getDeprecationReason = (target: any, key: string) => {
  return Reflect.getMetadata(GRAPHENE_DEPRECATED_KEY, target, key);
};

// Setup the decorators

// Description decorator.
// This decorator permits us to decorate classes, like:
//
//   @description('The class description')
//   class X {}
//
// Or the class methods, like:
//
//   class X{
//     @description('The method description')
//     myMethod() {}
//   }
export const description = (description: string) => (
  target: any,
  key?: string,
  descriptor?: PropertyDescriptor
) => {
  if (typeof key !== 'undefined') {
    // It's a decorated method
    setDescription(target, key, description);
    return descriptor;
  }
  // It's a decorated class
  setDescription(target, description);
  return target;
};

// This decorator permits us to mark fields as deprecated, like:
//
//   class X{
//     @deprecated('This method is deprecated')
//     myMethod() {}
//   }
export const deprecated = (reason: string) => (
  target: any,
  key: string,
  descriptor: PropertyDescriptor
) => {
  if (typeof key !== 'undefined') {
    // It's a decorated method
    // console.log(reason, target, key);
    setDeprecationReason(target, key, reason);
    return descriptor;
  }
  // It's a decorated class
  throw new Error(
    `Classes can't be decorated with the @deprecated decorator. Received ${
      target
    }.`
  );
};
