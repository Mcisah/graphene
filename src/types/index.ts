// import { getGraphQLType } from './../types_old/base';
import {
  GraphQLFieldConfig,
  GraphQLObjectType,
  // GraphQLArgumentConfig,
  isOutputType,
  // GraphQLInputType,
  isInputType,
  GraphQLArgumentConfig,
  graphql,
  printSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  defaultFieldResolver,
  GraphQLInputType,
  GraphQLInterfaceType,
  GraphQLTypeResolver,
  GraphQLID,
  GraphQLEnumType,
  GraphQLEnumValueConfigMap,
  GraphQLList,
  GraphQLNonNull
} from 'graphql';

import 'reflect-metadata';
import { GraphQLSchema } from 'graphql/type/schema';
import { getGraphQLType, setGraphQLType, setupNativeTypes } from './reflection';

export const GRAPHENE_FIELDS_METADATA_KEY = 'graphene:fields';
import { getDeprecationReason, getDescription } from './reflection';

export {
  description,
  getDescription,
  deprecated,
  getDeprecationReason
} from './reflection';

export const ID = GraphQLID;

setupNativeTypes();

export const Argument = (
  type: InputType,
  description?: string,
  defaultValue?: any
): ArgumentType => {
  return {
    type,
    description,
    defaultValue
  };
};

export const List = (ofType: any) => {
  return new GraphQLList(getGraphQLType(ofType));
};

export const NonNull = (ofType: any) => {
  return new GraphQLNonNull(getGraphQLType(ofType));
};

export type ArgumentType = {
  type: InputType;
  description?: string;
  defaultValue?: any;
};

type InputType =
  | GraphQLInputType
  | any
  | typeof String
  | typeof Number
  | typeof Boolean;
type FieldOptions = {
  args?: {
    [key: string]: ArgumentType | InputType;
  };
  description?: string;
  deprecationReason?: string;
};

type ArgumentMap = {
  [key: string]: GraphQLArgumentConfig;
};

type FieldType = (
  type?: any,
  options?: FieldOptions
) => (target: any, key: string) => void;

// type InputFieldType = (
//   type?: any,
//   options?: InputFieldOptions
// ) => (target: any, key: string) => void;

// type InputFieldOptions = {
//   description?: string;
//   defaultValue?: any;
// };

export const Field: FieldType = (type?: any, options: FieldOptions = {}) => (
  target: any,
  key: string
) => {
  // property value
  // console.log('Field', target, key);
  // console.log(type, target, key, target.constructor);

  // var _val = target[key];
  var _class = target.constructor;
  var fields: {
    [key: string]: () => GraphQLFieldConfig<any, any>;
  };
  if (!Reflect.hasMetadata(GRAPHENE_FIELDS_METADATA_KEY, _class)) {
    fields = {};
    Reflect.defineMetadata(GRAPHENE_FIELDS_METADATA_KEY, fields, _class);
  } else {
    fields = Reflect.getMetadata(GRAPHENE_FIELDS_METADATA_KEY, _class);
  }
  if (key in fields) {
    throw new Error(`Field ${key} is already defined.`);
  }
  fields[key] = () => {
    var _type = getGraphQLType(type);
    if (!isOutputType(_type)) {
      throw new Error('Type is not output');
    }
    // Arg construction
    // var decoratedArguments: ArgumentMap | null = null;
    // if (target[key]) {
    //   decoratedArguments = Reflect.getMetadata(
    //     GRAPHENE_ARGUMENTS_METADATA_KEY,
    //     target[key]
    //   );
    // }
    // console.log('FIELD', target, key, decoratedArguments);
    var argKey: string;
    var args = options.args || {};
    var fieldArgs: ArgumentMap = {};
    for (argKey in args) {
      var arg: ArgumentType | InputType = args[argKey];
      var extra: {};
      var argType: any;
      if (
        typeof (<ArgumentType>arg).type !== 'undefined' &&
        !isInputType(<GraphQLInputType>arg)
      ) {
        extra = {
          description: (<ArgumentType>arg).description
        };
        argType = (<ArgumentType>arg).type;
      } else {
        extra = {};
        argType = arg;
      }

      var newType = getGraphQLType(argType);
      if (!isInputType(newType)) {
        throw new Error(
          `Field argument ${argKey} expected to be Input type. Received: ${
            argType
          }.`
        );
      }
      fieldArgs[argKey] = {
        type: newType,
        ...extra
      };
    }
    var targetResolver = target[key];
    var resolver: GraphQLFieldResolver<any, any> = defaultFieldResolver;
    if (typeof targetResolver === 'function') {
      resolver = (
        root: any,
        args: { [argName: string]: any },
        context: any,
        info: GraphQLResolveInfo
      ) => {
        return targetResolver.call(root, args, context, info);
      };
    }
    return {
      args: fieldArgs,
      type: _type,
      description: getDescription(target, key),
      resolve: resolver
    };
  };
  // console.log('TYPE', Reflect.getMetadata('design:type', target, key));
  // console.log(
  //   'RETURN TYPE',
  //   Reflect.getMetadata('design:returntype', target, key)
  // );
};

// export const argument = (type?: any, options?: ArgumentOptions) => (
//   target: any,
//   key: string
//   // index: number
// ) => {
//   // console.log('argument', type, 'x', target, 'o', key, 'fin');
//   var baseArguments: ArgumentMap;
//   if (!Reflect.hasMetadata(GRAPHENE_ARGUMENTS_METADATA_KEY, target[key])) {
//     baseArguments = {};
//     Reflect.defineMetadata(
//       GRAPHENE_ARGUMENTS_METADATA_KEY,
//       baseArguments,
//       target[key]
//     );
//   } else {
//     baseArguments = Reflect.getMetadata(
//       GRAPHENE_ARGUMENTS_METADATA_KEY,
//       target[key]
//     );
//   }
//   baseArguments[key] = new Argument(type, options);
// };

export type ObjectTypeOptions = {
  name?: string;
  description?: string;
  interfaces?: any[];
};

// class MYI {
//   [key: string]: any;
// }

export const ObjectType = (opts: ObjectTypeOptions = {}) => <
  T extends { new (...args: any[]): any }
>(
  target: T
): T => {
  // save a reference to the original constructor
  type FieldsMap = {
    [key: string]: () => GraphQLFieldConfig<any, any>;
  };

  var interfaces: GraphQLInterfaceType[] = (opts.interfaces || []).map(
    iface => {
      var ifaceType = getGraphQLType(iface);
      if (!(ifaceType instanceof GraphQLInterfaceType)) {
        throw new Error('Provided interface is not valid');
      }
      return ifaceType;
    }
  );

  var allInterfaceFields: {
    [key: string]: () => GraphQLFieldConfig<any, any>;
  } = {};
  (opts.interfaces || []).forEach((_, index) => {
    var iface = (opts.interfaces || [])[index];
    var ifaceFields: FieldsMap = Reflect.getMetadata(
      GRAPHENE_FIELDS_METADATA_KEY,
      iface
    );
    if (!ifaceFields) {
      throw new Error(`Type ${target} must have at least one field.`);
    }
    allInterfaceFields = {
      ...allInterfaceFields,
      ...ifaceFields
    };
    // return ifaceFields;
  });

  var fields: {
    [key: string]: () => GraphQLFieldConfig<any, any>;
  } = {
    ...allInterfaceFields,
    ...(Reflect.getMetadata(GRAPHENE_FIELDS_METADATA_KEY, target) || {})
  };

  if (!Object.keys(fields).length) {
    throw new Error(`Type ${target} must have at least one field.`);
  }

  setGraphQLType(
    target,
    new GraphQLObjectType({
      name: opts.name || target.name,
      description: opts.description || getDescription(target),
      interfaces: interfaces,
      fields: () => {
        var key: string;
        var finalFields: {
          [key: string]: GraphQLFieldConfig<any, any>;
        } = {};
        // var ifaceFields: FieldsMap;
        // for (ifaceFields of interfacesFields) {
        //   for (key in ifaceFields) {
        //     finalFields[key] = ifaceFields[key]();
        //   }
        // }
        for (key in fields) {
          finalFields[key] = fields[key]();
        }
        return finalFields;
      }
    })
  );

  // var MYO = class extends target {
  //   [P in keyof target]?: target[P]
  // };
  return target;
  // return class extends target implements MYI {
  //   ra: boolean = true;
  // };

  // return new constructor (will override original)
  // return target;
};

export type InterfaceTypeOptions = {
  name?: string;
  description?: string;
  resolveType?: (root?: any, context?: any, info?: GraphQLResolveInfo) => any;
};

export const InterfaceType = (opts: InterfaceTypeOptions = {}) => <
  T extends { new (...args: any[]): any }
>(
  target: T
): T => {
  var fields: {
    [key: string]: () => GraphQLFieldConfig<any, any>;
  } = Reflect.getMetadata(GRAPHENE_FIELDS_METADATA_KEY, target);
  if (!fields) {
    throw new Error(`Type ${target} must have at least one field.`);
  }

  var resolveType: GraphQLTypeResolver<any, any> = (
    root?: any,
    context?: any,
    info?: GraphQLResolveInfo
  ): string | GraphQLObjectType | Promise<string | GraphQLObjectType> => {
    if (opts.resolveType) {
      root = opts.resolveType(root, context, info);
    }
    return <GraphQLObjectType>getGraphQLType(root);
  };

  setGraphQLType(
    target,
    new GraphQLInterfaceType({
      name: opts.name || target.name,
      description: opts.description || getDescription(target),
      resolveType: resolveType,
      fields: () => {
        var key: string;
        var finalFields: {
          [key: string]: GraphQLFieldConfig<any, any>;
        } = {};
        for (key in fields) {
          finalFields[key] = fields[key]();
        }
        return finalFields;
      }
    })
  );

  // var MYO = class extends target {
  //   [P in keyof target]?: target[P]
  // };
  return target;
  // return class extends target implements MYI {
  //   ra: boolean = true;
  // };

  // return new constructor (will override original)
  // return target;
};

class BaseClass {}

const BaseClassProperties = Object.getOwnPropertyNames(BaseClass);

// We remove the properties automatically included in the BaseClass
// Such as .length, .name and .prototype
const getStaticProperties = (_class: Object) => {
  return Object.getOwnPropertyNames(_class).filter(
    name => BaseClassProperties.indexOf(name) === -1
  );
};

type EnumOptions = {
  name?: string;
  description?: string;
};
export const EnumType = (opts: EnumOptions = {}) => <
  T extends { new (...args: any[]): {}; [key: string]: any }
>(
  target: T
): T => {
  var values: GraphQLEnumValueConfigMap = {};
  getStaticProperties(target).forEach(name => {
    values[name] = {
      value: target[name]
    };
  });
  setGraphQLType(
    target,
    new GraphQLEnumType({
      name: opts.name || target.name,
      description: opts.description || getDescription(target),
      values: values
    })
  );
  return target;
  // // return constructor;
  // return class extends constructor {};
};

type EnumValueOptions = {
  description?: string;
};

export const EnumValue = (options: EnumValueOptions = {}) => (
  target: any,
  key: string
) => {
  console.log('enumvalue', options, target, key);
};

export type SchemaOptions = {
  query: any;
};

export class Schema extends GraphQLSchema {
  constructor(options: SchemaOptions) {
    var queryType: GraphQLObjectType = <GraphQLObjectType>getGraphQLType(
      options.query
    );
    super({
      query: queryType
    });
  }
  execute(query: string, ...args: any[]) {
    return graphql(this, query, ...args);
  }
  toString() {
    return printSchema(this);
  }
}
