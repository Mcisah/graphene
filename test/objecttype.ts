import { getGraphQLType } from './../src/types/reflection';
import { ObjectType, Field, InterfaceType } from './../src/types/index';
import { GraphQLObjectType, GraphQLInterfaceType } from 'graphql';

describe('ObjectType setup', () => {
  test(`create ObjectType properly`, () => {
    @ObjectType()
    class MyObjectType {
      @Field(String) hello: string;
    }

    var graphqlType: GraphQLObjectType = <GraphQLObjectType>getGraphQLType(
      MyObjectType
    );
    expect(graphqlType.name).toBe('MyObjectType');
    expect(graphqlType.description).toBe(undefined);
  });
  test(`create ObjectType properly`, () => {
    @ObjectType({
      name: 'MyCustomObjectType',
      description: 'My Description'
    })
    class MyObjectType {
      @Field(String)
      hello(): string {
        return 'World';
      }
    }
    var graphqlType: GraphQLObjectType = <GraphQLObjectType>getGraphQLType(
      MyObjectType
    );

    expect(graphqlType.name).toBe('MyCustomObjectType');
    expect(graphqlType.description).toBe('My Description');
    expect(graphqlType.getFields()).toHaveProperty('hello');

    // We preserve the function
    expect(new MyObjectType().hello()).toBe('World');
  });

  test(`create ObjectType with interfaces`, () => {
    @InterfaceType()
    class MyInterface1 {
      @Field(String)
      foo1?(): string {
        return 'Foo1';
      }
    }

    @InterfaceType()
    class MyInterface2 {
      @Field(String)
      foo2?(): string {
        return 'Foo1';
      }
    }

    @ObjectType({
      interfaces: [MyInterface1, MyInterface2]
    })
    class MyObjectType {
      @Field(String)
      bar?(): string {
        return 'Bar';
      }
    }

    var graphqlType: GraphQLObjectType = <GraphQLObjectType>getGraphQLType(
      MyObjectType
    );

    expect(graphqlType.name).toBe('MyObjectType');

    var graphqlInterfaceType1: GraphQLInterfaceType = <GraphQLInterfaceType>getGraphQLType(
      MyInterface1
    );
    var graphqlInterfaceType2: GraphQLInterfaceType = <GraphQLInterfaceType>getGraphQLType(
      MyInterface2
    );

    expect(graphqlType.getInterfaces()).toEqual([
      graphqlInterfaceType1,
      graphqlInterfaceType2
    ]);

    expect(Object.keys(graphqlType.getFields())).toEqual([
      'foo1',
      'foo2',
      'bar'
    ]);
  });
});
