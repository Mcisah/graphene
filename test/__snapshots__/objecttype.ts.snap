// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectType setup create ObjectType properly 1`] = `
"# My Description
type MyCustomObjectType {
  hello: String
}"
`;
